### PR TITLE
Pytesting

### PR DESCRIPTION
**Pull request by harrisonpim**
Proof of concept for running relevance tests against elasticsearch indices with [pytest parametrised tests](https://docs.pytest.org/en/7.1.x/example/parametrize.html).

You should be able to run the tests locally with `poetry run rank test`

### CODE REVIEW

File: cli/secrets.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import boto3
+
+
+def get_secret(secret_id, region_name="eu-west-1"):
+    sts_client = boto3.client("sts")
```
Comment (jamieparkinson) [inline cli/secrets.py:5]: We should be creating 1 session per invocation of the whole CLI - pattern elsewhere is that functions which use AWS services have signatures like 
```python
def get_secret(session, *, secret_id)
```

File: pyproject.toml [hunk 2]
```diff
@@ -13,9 +13,14 @@ rank = "cli.main:app"
...
 python = "^3.10"
 typer = {extras = ["all"], version = "^0.9.0"}
+pytest = "^7.4.0"
+elasticsearch = "^8.9.0"
+boto3 = "^1.28.16"
+httpx = "^0.24.1"
 
 [tool.poetry.group.dev.dependencies]
 black = {extras = ["d"], version = "^23.7.0"}
+isort = "^5.12.0"
```
Comment (jamieparkinson) [inline pyproject.toml:23]: Think you also need to tell isort not to interfere with black https://pycqa.github.io/isort/docs/configuration/black_compatibility.html

File: cli/relevance_tests/test_precision.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+import pytest
+
+from . import get_pipeline_elastic_client, works_index, works_query
+
+client = get_pipeline_elastic_client()
+
+test_cases = [
+    pytest.param("horse", "pgwnkf2h", id="horse"),
+    pytest.param("cow", "wm8wy47d", id="cow"),
+]
```
Comment (jamieparkinson) [inline cli/relevance_tests/test_precision.py:12]: Aware that this is not a proper implementation but it's very unclear what these params are - I know this is partly pytest's fault, might be clearer to define the test cases inline in the `@pytest.mark.parametrize`?

**Review by jamieparkinson — APPROVED** [review]

File: cli/relevance_tests/test_precision.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+
+@pytest.mark.parametrize("search_terms, expected_id", test_cases)
+def test_precision(search_terms, expected_id):
+    populated_query_string = works_query.replace("{{query}}", search_terms)
+    populated_query = json.loads(populated_query_string)
+    response = client.search(index=works_index, query=populated_query)
+    first_id = response["hits"]["hits"][0]["_id"]
+    if first_id != expected_id:
+        pytest.fail(f"Expected {expected_id}, got {first_id}", pytrace=False)
```
[the remark this answers is not in the document]
Comment (harrisonpim) [reply cli/relevance_tests/test_precision.py:22]: It's definitely not the cleanest/clearest in its current state, but I've left this here as a hint towards what we might do later on. 

In cases where tests don't pass, but are not so catastrophic that we want to fail the whole suite, we can use `pytest.warns()` etc. 

IMO the flexibility of the output messages is also useful when tests can fail for multiple reasons, compared to plain asserts

**Pull request merged**